src/functions/BaselineCorrection.C

```diff
@@ -0,0 +1,247 @@
+
+// This code was implemented in collaboration with Christopher J. Wong
+// (chris.wong@utah.edu) from the University of Utah.
+
+// MASTODON includes
+#include "BaselineCorrection.h"
+#include "BaselineCorrectionUtils.h"
+
+// MOOSE includes
+#include "DelimitedFileReader.h"
+
+registerMooseObject("MastodonApp", BaselineCorrection);
+
+InputParameters
+BaselineCorrection::validParams()
+{
+  InputParameters params = Function::validParams();
+  params.addClassDescription("Applies a baseline correction to an accceleration time history "
+                             "using least squares polynomial fits and outputs the adjusted "
+                             "acceleration with linear interpolation.");
+
+  params.addParam<FileName>(
+      "data_file", "The name of a CSV file containing raw acceleration time history data.");
+  params.addParam<std::string>(
+      "time_name",
+      "The header name of the column which contains the time values in the data file. If not "
+      "specified, they are assumed to be in the first column index.");
+  params.addParam<std::string>(
+      "acceleration_name",
+      "The header name for the column which contains the acceleration values in the data file. If "
+      "not specified, they are assumed to be in the second column index.");
+
+  params.addParam<std::vector<Real>>("time_values", "The time abscissa values.");
+  params.addParam<std::vector<Real>>("acceleration_values", "The acceleration ordinate values.");
+
+  params.addRequiredParam<Real>("gamma", "The gamma parameter for Newmark time integration.");
+  params.addRequiredParam<Real>("beta", "The beta parameter for Newmark time integration.");
+
+  params.addRangeCheckedParam<unsigned int>(
+      "accel_fit_order",
+      "(0 <= accel_fit_order) & (accel_fit_order < 10)",
+      "If this is provided, the acceleration time history will be adjusted using an nth-order "
+      "polynomial fit of the nominal acceleration data, where n = accel_fit_order (only integer "
+      "values from 0 to 9 are supported).");
+  params.addRangeCheckedParam<unsigned int>(
+      "vel_fit_order",
+      "(0 <= vel_fit_order) & (vel_fit_order < 10)",
+      "If this is provided, the acceleration time history will be adjusted using an nth-order "
+      "polynomial fit of the nominal velocity data, where n = vel_fit_order (only integer values "
+      "from 0 to 9 are supported).");
+  params.addRangeCheckedParam<unsigned int>(
+      "disp_fit_order",
+      "(0 <= disp_fit_order) & (disp_fit_order < 10)",
+      "If this is provided, the acceleration time history will be adjusted using an nth-order "
+      "polynomial fit of the nominal displacement data, where n = disp_fit_order (only integer "
+      "values from 0 to 9 are supported).");
+
+  params.addParam<Real>("scale_factor",
+                        1.0,
+                        "A scale factor to be applied to the adjusted acceleration time history.");
+  params.declareControllable("scale_factor");
```

Comment:
Is this for the stochastic simulations stuff? If so, cool!

Reply:
The scale factor is mostly for whatever purpose. Most MOOSE `Function` objects do this.